### PR TITLE
Add undefined to Symbol.valueDeclaration

### PR DESCRIPTION
**Pull request by sandersn**
Also add undefined to a number of utility functions that have always accepted it, but never added it to their type.

### CODE REVIEW

File: tests/baselines/reference/api/tsserverlibrary.d.ts [hunk 1]
```diff
@@ -4107,7 +4107,7 @@ declare namespace ts {
     function idText(identifierOrPrivateName: Identifier | PrivateIdentifier): string;
     function symbolName(symbol: Symbol): string;
     function getNameOfJSDocTypedef(declaration: JSDocTypedefTag): Identifier | PrivateIdentifier | undefined;
-    function getNameOfDeclaration(declaration: Declaration | Expression): DeclarationName | undefined;
+    function getNameOfDeclaration(declaration: Declaration | Expression | undefined): DeclarationName | undefined;
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/api/tsserverlibrary.d.ts:4110]: Not sure how I feel about changing the public API on these, but maybe it's fine.

File: src/services/codefixes/inferFromUsage.ts [hunk 2]
```diff
@@ -1101,15 +1101,16 @@ namespace ts.codefix {
                 const genericParam = genericSig.parameters[i];
                 const usageParam = usageSig.parameters[i];
                 const isRest = genericSig.declaration && isRestParameter(genericSig.declaration.parameters[i]);
-                if (!usageParam) {
+                if (!usageParam/* || !usageParam.valueDeclaration || !genericParam.valueDeclaration*/) {
```
Comment (DanielRosenwasser) [inline src/services/codefixes/inferFromUsage.ts:1104]: Mm?

File: src/services/codefixes/inferFromUsage.ts [hunk 3]
```diff
@@ -950,7 +950,7 @@ namespace ts.codefix {
             const props = createMultiMap<Type>();
             for (const anon of anons) {
                 for (const p of checker.getPropertiesOfType(anon)) {
-                    props.add(p.name, checker.getTypeOfSymbolAtLocation(p, p.valueDeclaration));
+                    props.add(p.name, p.valueDeclaration ? checker.getTypeOfSymbolAtLocation(p, p.valueDeclaration!) : checker.getAnyType());
```
Comment (DanielRosenwasser) [inline src/services/codefixes/inferFromUsage.ts:953]:
```suggestion
                    props.add(p.name, p.valueDeclaration ? checker.getTypeOfSymbolAtLocation(p, p.valueDeclaration) : checker.getAnyType());
```

File: src/services/codefixes/fixSpelling.ts [hunk 4]
```diff
@@ -87,7 +87,7 @@ namespace ts.codefix {
         const suggestion = symbolName(suggestedSymbol);
         if (!isIdentifierText(suggestion, target) && isPropertyAccessExpression(node.parent)) {
             const valDecl = suggestedSymbol.valueDeclaration;
-            if (isNamedDeclaration(valDecl) && isPrivateIdentifier(valDecl.name)) {
+            if (valDecl &&  isNamedDeclaration(valDecl) && isPrivateIdentifier(valDecl.name)) {
```
Comment (DanielRosenwasser) [inline src/services/codefixes/fixSpelling.ts:90]:
```suggestion
            if (valDecl && isNamedDeclaration(valDecl) && isPrivateIdentifier(valDecl.name)) {
```

File: tests/baselines/reference/api/tsserverlibrary.d.ts (hunk 1 above)
Comment (sandersn) [reply tests/baselines/reference/api/tsserverlibrary.d.ts:4110]: I think it's fine: they just make the function easier to call, and they're on functions that already return `undefined` or `boolean`, which implies (correctly) that they pass through `undefined` if you call them.

File: src/services/codefixes/inferFromUsage.ts (hunk 2 above)
Comment (sandersn) [reply src/services/codefixes/inferFromUsage.ts:1104]: Leftover from my first attempt, same as the extraneous `!` above.

**Review by andrewbranch — APPROVED** [review]

**Review by sheetalkamat — APPROVED** [review]

**Review by weswigham — APPROVED** [review]

**Pull request merged**